<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Datasets &mdash; chemicalchecker 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="assets/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="assets/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="assets/css/logo.css?v=fc604eda" />
      <link rel="stylesheet" type="text/css" href="assets/css/overrides.css?v=466af0f7" />

  
    <link rel="shortcut icon" href="assets/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="assets/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="assets/jquery.js?v=5d32c60e"></script>
        <script src="assets/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="assets/documentation_options.js?v=8d563738"></script>
        <script src="assets/doctools.js?v=888ff710"></script>
        <script src="assets/sphinx_highlight.js?v=dc90522c"></script>
    <script src="assets/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Signaturization" href="signaturization.html" />
    <link rel="prev" title="Chemical Checker Manifesto" href="manifesto.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="assets/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="manifesto.html">Chemical Checker Manifesto</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Datasets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#levels-coordinates-and-datasets">Levels, coordinates and datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#dataset-characteristics">Dataset characteristics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#dataset-pre-processing">Dataset pre-processing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="signaturization.html">Signaturization</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">chemicalchecker</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Datasets</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/sources.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="datasets">
<h1>Datasets<a class="headerlink" href="#datasets" title="Link to this heading"></a></h1>
<p>In the CC nomenclature, a dataset is determined by:</p>
<ol class="arabic simple">
<li><p>One coordinate.</p></li>
<li><p>One (typically) or multiple (eventually) sources having the same type
of (mergeable) data.</p></li>
<li><p>A <cite>pre-processing procedure</cite> yielding signatures type 0.</p></li>
</ol>
<section id="levels-coordinates-and-datasets">
<h2>Levels, coordinates and datasets<a class="headerlink" href="#levels-coordinates-and-datasets" title="Link to this heading"></a></h2>
<p>The CC is divided into five <strong>levels</strong> of increasing complexity:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">A</span></code></p></td>
<td><p>Chemistry</p></td>
<td><p>Chemical properties of the compounds.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">B</span></code></p></td>
<td><p>Targets</p></td>
<td><p>Chemical-protein interactions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">C</span></code></p></td>
<td><p>Networks</p></td>
<td><p>Higher-order effects of small molecules.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">D</span></code></p></td>
<td><p>Cells</p></td>
<td><p>Readouts of compound cell-based assays.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">C</span></code></p></td>
<td><p>Clinics</p></td>
<td><p>Clinical data of drugs and environmental chemicals.</p></td>
</tr>
</tbody>
</table>
<p>In turn, each level is divided into 5 sublevels or <strong>coordinates</strong>
representing different aspects of the data. Each sublevel has an
<em>exemplary</em> dataset, as described below:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Coordinate</p></th>
<th class="head"><p>Name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">A1</span></code></p></td>
<td><p>2D fingerprints</p></td>
<td><p>Binary representation
of the 2D structure
of a molecule. The
neighbourhood of
every atom is encoded
using circular
topology hashing.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">A2</span></code></p></td>
<td><p>3D fingerprints</p></td>
<td><p>Similar to <code class="docutils literal notranslate"><span class="pre">A1</span></code>,
the 3D structures of
the three best
conformers after
energy minimization
are hashed into a
binary representation
without the need for
structural alignment.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">A3</span></code></p></td>
<td><p>Scaffolds</p></td>
<td><p>Largest molecular
scaffold (usually a
ring system)
remaining after
applying Murcko’s
pruning rules.
Additionally, we keep
the corresponding
framework, i.e. a
version of the
scaffold where all
atoms are carbons and
all bonds are single.
The scaffold and the
framework are encoded
with path-based
1024-bit
fingerprints,
suitable for
capturing
substructures in
similarity searches.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">A4</span></code></p></td>
<td><p>Structural keys</p></td>
<td><p>166 functional groups
and substructures
widely accepted by
medicinal chemists
(MACCS keys).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">A5</span></code></p></td>
<td><p>Physicochemistry</p></td>
<td><p>Physicochemical
properties such as
molecular weight,
logP, and
refractivity. Number
of hydrogen-bond
donors and acceptors,
rings, etc.
Drug-likeness
measurements
e.g. number of
structural alerts,
Lipinski’s rule-of-5
violations or
chemical beauty
(QED).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">B1</span></code></p></td>
<td><p>Mechanism of action</p></td>
<td><p>Drug targets with
known pharmacological
action and modes
(agonist, antagonist,
etc.).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">B2</span></code></p></td>
<td><p>Metabolic genes</p></td>
<td><p>Drug metabolizing
enzymes,
transporters, and
carriers.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">B3</span></code></p></td>
<td><p>Crystals</p></td>
<td><p>Small molecules
co-crystalized with
protein chains. Data
is organized
according to the
structural families
of the protein
chains.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">B4</span></code></p></td>
<td><p>Binding</p></td>
<td><p>Compound–protein
binding data
available in major
public chemogenomics
databases. Data
mainly comes from
academic publications
and patents. Only
binding affinities
below a
class-specific
threshold are kept
(kinases ≤ 30 nM,
GPCRs ≤ 100 nM,
nuclear receptors ≤
100 nM, ion channels
≤ 10 uM and others ≤
1 uM).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">B5</span></code></p></td>
<td><p>HTS bioassays</p></td>
<td><p>Hits from screening
campaigns against
protein targets
(mainly confirmatory
functional assays
below 10 uM).</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">C1</span></code></p></td>
<td><p>Biological roles</p></td>
<td><p>Ontology terms
associated with small
molecules with
recognized biological
roles, such as known
drugs, metabolites
and other natural
products.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">C2</span></code></p></td>
<td><p>Metabolic network</p></td>
<td><p>Curated
reconstruction of
human metabolism,
containing
metabolites and
reactions. Data is
represented as a
network where nodes
are metabolites and
edges connect
substrates and
products of
reactions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">C3</span></code></p></td>
<td><p>Canonical pathways</p></td>
<td><p>Canonical pathways
related to the known
receptors of
compounds (as
recorded in <code class="docutils literal notranslate"><span class="pre">B4</span></code>).
Pathways are assigned
via a
guilt-by-association
approach, i.e. a
molecule is related
to a pathway if at
least one of the
molecule targets is a
member of it.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">C4</span></code></p></td>
<td><p>Biological processes</p></td>
<td><p>Similar to <code class="docutils literal notranslate"><span class="pre">C3</span></code>,
biological processes
from the gene
ontology are
associated with
compounds via a
guilt-by-association
approach from <code class="docutils literal notranslate"><span class="pre">B4</span></code>
data. All parent
terms are kept, from
the leaves of the
ontology to its root.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">C5</span></code></p></td>
<td><p>Interactomes</p></td>
<td><p>Neighborhoods of
<code class="docutils literal notranslate"><span class="pre">B4</span></code> targets are
collected by
inspecting several
large protein-protein
interaction networks.
A random-walk
algorithm is used to
obtain a robust
measure of
‘proximity’ in the
network.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">D1</span></code></p></td>
<td><p>Gene expression</p></td>
<td><p>Transcriptional
response of cell
lines upon exposure
to small molecules. A
well-documented
reference dataset of
gene expression
profiles is used to
map all compound
profiles using a
two-sided gene set
enrichment analysis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">D2</span></code></p></td>
<td><p>Cancer cell lines</p></td>
<td><p>Small molecule
sensitivity data
(GI50) of a panel of
60 cancer cell lines.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">D3</span></code></p></td>
<td><p>Chemical genetics</p></td>
<td><p>Growth inhibition
profiles in a panel
of ~300 yeast
mutants. Data are
combined with yeast
genetic interaction
data so that
compounds can be
assimilated to
genetic alterations
when they have
similar profiles.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">D4</span></code></p></td>
<td><p>Morphology</p></td>
<td><p>Changes in U-2 OS
cell morphology
measured after
compound treatment
using a
mu
ltiplexed-cytological
cell painting assay.
812 morphology
features are recorded
via automated
microscopy and image
analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">D5</span></code></p></td>
<td><p>Cell bioassays</p></td>
<td><p>Small molecule cell
bioassays reported in
ChEMBL, mainly growth
and proliferation
measurements found in
the literature.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">E1</span></code></p></td>
<td><p>Therapeutic areas</p></td>
<td><p>Anatomical
Therapeutic Chemical
(ATC) codes of drugs.
All ATC levels are
considered.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">E2</span></code></p></td>
<td><p>Indications</p></td>
<td><p>Indications of
approved drugs and
drugs in clinical
trials. A controlled
medical vocabulary is
used.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">E3</span></code></p></td>
<td><p>Side effects</p></td>
<td><p>Side effects
extracted from drug
package inserts via
text-mining
techniques.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">E4</span></code></p></td>
<td><p>Disease phenotypes</p></td>
<td><p>Manually curated
relationships between
chemicals and
diseases. Chemicals
include drug
molecules and
environmental
substances, among
others.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">E5</span></code></p></td>
<td><p>Drug-drug
interactions</p></td>
<td><p>Changes in the effect
of a drug when it is
taken together with a
second drug.
Drug-drug
interactions may
alter
pharmacokinetics
and/or cause side
effects.</p></td>
</tr>
</tbody>
</table>
<p>Each of the coordinates can contain an arbitrary number of <strong>datasets</strong>
with increasing number (e.g. <code class="docutils literal notranslate"><span class="pre">A1.001</span></code>).</p>
</section>
<section id="dataset-characteristics">
<h2>Dataset characteristics<a class="headerlink" href="#dataset-characteristics" title="Link to this heading"></a></h2>
<p>This is how we define a dataset:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Values</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Code</p></td>
<td><p>e.g.<code class="docutils literal notranslate"><span class="pre">A1.001</span></code></p></td>
<td><p>Identifier of the
dataset.</p></td>
</tr>
<tr class="row-odd"><td><p>Level</p></td>
<td><p>e.g. <code class="docutils literal notranslate"><span class="pre">A</span></code></p></td>
<td><p>The CC level.</p></td>
</tr>
<tr class="row-even"><td><p>Coordinate</p></td>
<td><p>e.g.<code class="docutils literal notranslate"><span class="pre">A1</span></code></p></td>
<td><p>Coordinates in the CC
organization.</p></td>
</tr>
<tr class="row-odd"><td><p>Name</p></td>
<td><p>2D fingerprints</p></td>
<td><p>Display, short-name
of the dataset.</p></td>
</tr>
<tr class="row-even"><td><p>Technical name</p></td>
<td><p>1024-bit Morgan
fingerprints</p></td>
<td><p>A more technical name
for the dataset,
suitable for
chemo
-/bio-informaticians.</p></td>
</tr>
<tr class="row-odd"><td><p>Description</p></td>
<td><p>2D fingerprints are…</p></td>
<td><p>This field contains a
long description of
the dataset. It is
important that the
curator outlines here
the importance of the
dataset, why did
he/she make the
decision to include
it, and what are the
scenarios where this
dataset may be
useful.</p></td>
</tr>
<tr class="row-even"><td><p>Unknowns</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code>/<code class="docutils literal notranslate"><span class="pre">False</span></code></p></td>
<td><p>Does the dataset
contain known/unknown
data? Binding data
from chemogenomics
datasets, for
example, are
positive-unlabeled,
so they do contain
unknowns. Conversely,
chemical fingerprints
or gene expression
data do not contain
unknowns.</p></td>
</tr>
<tr class="row-odd"><td><p>Discrete</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code>/<code class="docutils literal notranslate"><span class="pre">False</span></code></p></td>
<td><p>The type of data that
ultimately expresses
de dataset, after the
pre-processing.
Categorical variables
are not allowed; they
must be converted to
one-hot encoding or
binarized. Mixed
variables are not
allowed, either.</p></td>
</tr>
<tr class="row-even"><td><p>Keys</p></td>
<td><p>e.g. <code class="docutils literal notranslate"><span class="pre">CPD</span></code> (we use
&#64;afernandez
<code class="docutils literal notranslate"><span class="pre">Bioteque</span></code>
nomenclature). Can be
<code class="docutils literal notranslate"><span class="pre">NULL</span></code>.</p></td>
<td><p>In the core CC
database, most of the
times this field will
correspond to
<code class="docutils literal notranslate"><span class="pre">CPD</span></code>, as the CC is
centred on small
molecules. It only
makes sense to have
keys of different
types when we do
connectivity
attempts, that is,
for example, when
mapping disease gene
expression
signatures.</p></td>
</tr>
<tr class="row-odd"><td><p>Features</p></td>
<td><p>e.g. <code class="docutils literal notranslate"><span class="pre">GEN</span></code> (we use
<code class="docutils literal notranslate"><span class="pre">Bioteque</span></code>
nomenclature). Can be
<code class="docutils literal notranslate"><span class="pre">NULL</span></code>.</p></td>
<td><p>When features
correspond to
explicit knowledge,
such as proteins,
gene ontology
processes, or
indications, we
express with this
field the type of
biological entities.
It is not allowed to
mix different feature
types. Features can,
however, have no
type, typically when
they come from a
heavily-processed
dataset, such as
gene-expression data.
Even if we use
<code class="docutils literal notranslate"><span class="pre">Bioteque</span></code>
nomenclature to the
define the type of
biological data, it
is not mandatory that
the vocabularies are
the ones used by the
<code class="docutils literal notranslate"><span class="pre">Bioteque</span></code>; for
example, I can use
non-human UniProt
ACs, if I deem it
necessary.</p></td>
</tr>
<tr class="row-even"><td><p>Exemplary</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code>/<code class="docutils literal notranslate"><span class="pre">False</span></code></p></td>
<td><p>Is the dataset
exemplary of the
coordinate. Only one
exemplary dataset is
valid for each
coordinate. Exemplary
datasets should have
good coverage (both
in keys space and
feature space) and
acceptable quality of
the data.</p></td>
</tr>
<tr class="row-odd"><td><p>Public</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code>/<code class="docutils literal notranslate"><span class="pre">False</span></code></p></td>
<td><p>Some datasets are
public, and some are
not, especially those
that come from
collaborations with
the pharma industry.</p></td>
</tr>
</tbody>
</table>
<p>See the <code class="xref py py-mod docutils literal notranslate"><span class="pre">chemicalchecker.database</span></code> for more information.</p>
</section>
<section id="dataset-pre-processing">
<h2>Dataset pre-processing<a class="headerlink" href="#dataset-pre-processing" title="Link to this heading"></a></h2>
<p>Dataset pre-processing refers to everything that happens from
downloaded/calculated/user-defined data until Signature Type 0.
Pre-processing can be of very different complexity:</p>
<img alt="assets/images/preprocessing.png" src="assets/images/preprocessing.png" />
<p>Here is where most of the SB&amp;NB research happens. For now, dataset
pre-processing is organized in a rather independent structure, i.e. each
dataset receives its pre-processing scripts
(see <code class="xref py py-mod docutils literal notranslate"><span class="pre">chemicalchecker.core.preprocess</span></code>).</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="manifesto.html" class="btn btn-neutral float-left" title="Chemical Checker Manifesto" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="signaturization.html" class="btn btn-neutral float-right" title="Signaturization" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, SBNB.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>
